### PR TITLE
chore: create ts/nextjs/express shell

### PR DESCRIPTION
**Pull request by portenez**
This is intended to be merged after #2 

**Describe your changes**
Init minimal project shell and linters (no ci actions for now)

**Testing performed**

- `npm run dev` works locally
- `npm eslint` works locally
- `npm stylelint` works locally
- `npm test` works locally

**Additional context**
There will be another PR for CI automation for the linting and unit tests

### CODE REVIEW

File: .eslintignore
Comment (dakotalillie) [inline .eslintignore:2]: Are we using [storybook](https://storybook.js.org)?

File: package.json
Comment (dakotalillie) [inline package.json]: At this point, I'm unsure if `coffechat` is just a typo that has popped up repeatedly, or it's actually the name of the project 😜

File: .eslintrc.yml
Comment (dakotalillie) [inline .eslintrc.yml]: I would personally prefer not to use nested ternaries. Do you find them useful?

File: .eslintrc.yml
Comment (dakotalillie) [inline .eslintrc.yml]: I'm curious why we need formatting rules like this for eslint. I'd imagine they'd fall under the purview of prettier.

```suggestion
    - avoidEscape: true # has to always be true to not clash with prettier
```

File: .gitignore
Comment (dakotalillie) [inline .gitignore:86]: **Nitpick**

Most of what's in this `.gitignore` does not seem to be used in this project.

File: .prettierrc.yml
Comment (dakotalillie) [inline .prettierrc.yml]: Willing to add `proseWrap: always`, by any chance? And maybe bump `printWidth` up to 100 or 120?

File: README.md
Comment (dakotalillie) [inline README.md]: Why do we need to run `npm init`? This project already has a `package.json`

File: README.md
Comment (dakotalillie) [inline README.md]: Perhaps worth adding a note explaining more thoroughly what this does?

File: doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md
Comment (dakotalillie) [inline doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md]:
```suggestion
- [React.js](https://reactjs.org/) is well-known and widely used `js` library
```

File: doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md
Comment (dakotalillie) [inline doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md]: I might prefer [Emotion](https://emotion.sh/docs/introduction) to Styled Components — they have very similar APIs at this point, but I've found Emotion to be a bit more reliable and easier to set up. What do you think?

File: doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md
Comment (dakotalillie) [inline doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md]:
```suggestion
- [eslint](https://eslint.org/) is a static analysis tool to find bugs
```

File: doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md
Comment (dakotalillie) [inline doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md]:
```suggestion
  conventions in your styles, and works with styled-components out of the box.
```

File: doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md
Comment (dakotalillie) [inline doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md]:
```suggestion
  spending valuable development time agreeing on a code style.
```

File: package.json
Comment (dakotalillie) [inline package.json]: Could probably use a catchier description here. @pooja-malpani do you have anything on hand?

File: package.json
Comment (dakotalillie) [inline package.json]: I'd imagine we probably need `prettier` and `prettier:fix` commands here as well, no?

File: src/pages/_document.tsx
Comment (dakotalillie) [inline src/pages/_document.tsx]: Correct me if I'm wrong, but since we're using React 17, I don't believe it's necessary to `import React from 'react'` anymore, here or elsewhere.

File: src/lib/Config.ts
Comment (dakotalillie) [inline src/lib/Config.ts]: Why is this capitalized, if it's not a class, component, or factory function?

File: src/lib/Config.ts
Comment (dakotalillie) [inline src/lib/Config.ts]: Personally, I find this file rather confusing. It seems like it is parsing, validating, and providing overrides and default values all at once. Am I understanding it correctly? If so, what would you say to splitting those things up into separate, single-purpose functions which can be chained?

File: src/pages/_document.tsx
Comment (dakotalillie) [inline src/pages/_document.tsx]: Could we use a better name for this than `MyDocument`, or add a comment to indicate why a custom `Document` is necessary?

File: tsconfig.json
Comment (dakotalillie) [inline tsconfig.json]: What is `typings`?

File: tsconfig.server.json
Comment (dakotalillie) [inline tsconfig.server.json]: Would esnext work here?

```suggestion
    "lib": ["esnext"],
    "target": "esnext",
```

**Review by dakotalillie — DISMISSED** [review]

File: .eslintignore
Comment (portenez) [reply .eslintignore:2]: Not yet, but I was planning to add it later. 

Removing this makes sense until we actually add storybook. However, leaving it is not harmful either. 

What's your pref?

File: package.json
Comment (portenez) [reply package.json]: Oh, man. It's a quirk of my Spanglish brain. Cafe/coffee merge in my mind, so I keep repeating the same mistake. I def, should fix.

File: .eslintrc.yml
Comment (portenez) [reply .eslintrc.yml]: Yeah, I do find them useful, but I see the harm. I'm totally ok with disallowing.

File: .gitignore
Comment (portenez) [reply .gitignore:86]: I used https://www.toptal.com/developers/gitignore to generate the `.gitignore` file.

Lately, I've been opting to do this because I've been burned by crafting the `.gitignore` manually. Sometimes, I forget (or don't even know I have to) add files to `.gitignore` and end up committing something that I shouldn't.

Lemme know what you think after my explanation. I'm fine with cleaning this up.

File: .prettierrc.yml
Comment (portenez) [reply .prettierrc.yml]: for `printWidth` my only problem is that I constantly split my editors vertically in `vscode`.  I also often put a terminal and vscode in the same monitor splitted vertically across the window 

- 120 is definitely too long and will give me a lot of hassle.
- 100 is a bit better, but still gives me some hassle. 

See screenshot:
image

File: package.json
Comment (portenez) [reply package.json]: So, the way this is set up, `esllint` will run prettier as part `eslint` and `eslint:fix`. So we're good on the javascript files.

For the markdown, yaml, etc. I was not planning/requiring on running prettier, but that's useful.  do you think I should do this?

File: src/pages/_document.tsx
Comment (portenez) [reply src/pages/_document.tsx]: Needed to update an eslint rule. Should be good now.

File: src/pages/_document.tsx
Comment (portenez) [reply src/pages/_document.tsx]: ah, yeah.  I copied it over from the docs. Will add a comment with a link.

As for the name, can you help me with a suggestion? I'm blanking on a good name: `CustomDoc` is as bad as `MyDocument` IMO. Then, there's only one single `_document.tsx` in next js, so we cannot make it super specific like `StyledComponentsBootstrapDoc` or something like that, because if we ever change it won't make sense.

File: tsconfig.json
Comment (portenez) [reply tsconfig.json]: ah, it's an old school way of defining the `typescript types`. There use to be competitor to `definitely` types used `typings` instead of `@types`. I'll remove

**Review by portenez — COMMENTED** [review]
I made some corrections and addressed some comments. I will try to address these ones soon. 

In other words, I wasn't able to go through all the comments in this pass.

Also, if anyone has any time to help me finalize this, please do help me. I mean add to this PR branch.

File: .eslintignore
Comment (dakotalillie) [reply .eslintignore:2]: I'm open to leaving it if we're eventually going to use it. My understanding is it's usually intended for component libraries — would we use it here to preview components in isolation before adding them to the frontend proper?

File: .gitignore
Comment (dakotalillie) [reply .gitignore:86]: Eh. It's a major nitpick. Ain't doin no harm.

File: .prettierrc.yml
Comment (dakotalillie) [reply .prettierrc.yml]: Fair enough — keeping `printWidth` at 80 is fine. Any chance for `proseWrap`?

Comment (dakotalillie) [reply .prettierrc.yml]: Ah sorry, I see now that you added it. Was using the wrong view.

File: package.json
Comment (dakotalillie) [reply package.json]: Gotcha. I personally like using Prettier on markdown in particular

**Comment by dakotalillie** [comment]
I'll see about pushing up some commits to address my own suggestions.

**Comment by portenez** [comment]
So I pulled the changes from master using the github ui, and that might create a messy commit history. Before merging let's see if we can cleanup the history (squash merge is disabled in this repo)

**Comment by dakotalillie** [comment]
> squash merge is disabled in this repo

😱

**Comment by portenez** [comment]
I'm in the car (wife driving). I resolved the storybook comment, and now can't get it back easily (in mobile ui).

But storybook is good for not only to preview but also for:

- writing simple snapshot tests at the same time the stories are written
- paralellize work. Devs can start working on components while backend work is still not ready.
- iterate with ux quickly (the storybook is easy to deploy to personal github pages)

**Comment by dakotalillie** [comment]
Righto. Sounds good to me 👍

File: src/pages/_document.tsx
Comment (dakotalillie) [reply src/pages/_document.tsx]: changed to `DocumentWithStyles`. Let me know if you'd prefer something different.

**Comment by portenez** [comment]
Thank you. Everything looks great. The only thing to double check is the target for tsconfig server. That one needs to match the nodejs version that we're targeting.

**Comment by portenez** [comment]
Whoever gets a chance, let's update the tsconfig version according to this
https://github.com/tsconfig/bases#centralized-recommendations-for-tsconfig-bases

File: src/lib/Config.ts
Comment (dakotalillie) [reply src/lib/Config.ts]: I rewrote this using [ajv](https://ajv.js.org)

File: src/server/env.ts
Comment (portenez) [inline src/server/env.ts:13]: very cool! So, pretty much I had reimplemented this functionality.

File: src/server/env.ts
Comment (portenez) [inline src/server/env.ts]: **nitpick**:

ok, this might be wishful thinking... but is there any way we could somehow export this in a way we could import like:

```typescript
import { PORT } from './env'
```

this is already pretty cool, but this would be very convenient.

File: src/server/index.ts
Comment (portenez) [inline src/server/index.ts]: Per my other comment: https://github.com/bloomberg/coffeechat/pull/3/commits/92b506e67f2b647376cfb85836ea3e17ae7bc480#r624785942



I'm wondering if we could somehow export the properties. This is already pretty, good. I'm just wondering.

File: src/server/env.ts
Comment (portenez) [inline src/server/env.ts]: ok, I'm just wondering.. `process.env` interface is actually `Record<string, string>`. Would this affect at all this implementation? Could this make the use of this a bit confusing?

Also, we might be able to change the types in the `makeParser` funct, to make it clearer.

File: src/lib/makeParser.spec.ts
Comment (portenez) [inline src/lib/makeParser.spec.ts]: I have no hard pref, but I tend to use `__test__/*.spec.ts` for the tests. What standard should we use?
with or without `__test__` ?

File: README.md
Comment (portenez) [reply README.md]: What do you have in mind? Maybe, add the note the way you think would be helpful?

File: src/server/env.ts
Comment (portenez) [inline src/server/env.ts]: @dakotalillie What do you think of this? it's a bit verbose... but gives me the convenience I was looking for lol. You're either verbose here, or later.

File: README.md
Comment (portenez) [inline README.md:32]: I ran prettier on everything. That's why the changes. please ignore whitespace for a better experience.

**Comment by portenez** [comment]
@dakotalillie  I pushed a few things:

- Better types for `makeParser`
- My wish of being able to use `import { PORT } from './env'
- `.vscode` with standardized formatting, recommended plugins and such (there are pros and cons to this, lemme know what do you think). At least the recommendations should be there. As for the settings, It's useful for people to immediately get started with a working config. 
- I ran prettier on everything and added prettier targets.

Let's try to merge this soon. 
As for the `emotion`, let's do that as a quick follow-up? I have used emotion before but never with nextjs. My pet peeve with emotion is having to do this for the global styles:
https://emotion.sh/docs/globals

Lol, obviously it's a nitpick. So, in sum yeah I'm good with emotion.

File: src/lib/makeParser.spec.ts
Comment (dakotalillie) [reply src/lib/makeParser.spec.ts]: I tend to like to put the spec files directly next to what they're testing, because in general I like to 1. colocate code as much as possible, and 2. use flat structures as much as possible. However, I do think it's just a matter of personal preference — as the project lead, if you'd prefer to use `__test__`, I'll defer to you.

File: src/server/env.ts
Comment (dakotalillie) [reply src/server/env.ts]: I think it's a good idea to destructure all the properties in here! Updated to make it a bit more concise, but I like the move in general.

**Comment by dakotalillie** [comment]
@portenez those type improvements do make it a lot cleaner — thank you for that.

I'm un-opinionated regarding adding the `.vscode` settings. I tend not to add them myself, but I am open-minded about it being a beneficial move which I just haven't explored yet. Just made a couple minor tweaks, including changing the extension for `settings.json` to `.jsonc` since you have a comment in there.

I've just been in a couple situations where Styled Components has inadvertently released a breaking change in a minor version bump, or where I struggled to get it working so I switched to Emotion which worked immediately. There were also some comparisons a while back about how Emotion was smaller and performed better than Styled Components, although I think that gap may have been bridged.

I think these are all nitpicks — I'm good with Styled Components too. Just wanted to bring up the point and have a small conversation. Ultimately, as with the conversation around using `__test__` directories, I'll defer to you.

**Comment by dakotalillie** [comment]
Changed the server target to ES2020 given the [Node Target Mappings](https://github.com/microsoft/TypeScript/wiki/Node-Target-Mapping) and the fact that the `.nvmrc` and `package.json` both say to use Node 14.

**Review by dakotalillie — DISMISSED** [review]
Feel free to make any additional changes based on our discussions around placement of tests and potential usage of Emotion. This all looks good to me 🎉

**Comment by portenez** [comment]
@dakotalillie Can you review/approve again? I did a few things and the approval was dismissed.

Here is what I did:

- removed the `merge` commit from `main` to get a (slightly) cleaner git history
- rebased from `main`
- added two commits:
  - Small updates to the README
  - Switched to using `__tests__` , and only configuring `jest` to only allow that pattern.

**Comment by portenez** [comment]
Also, I switched this to emotion! 🧑‍🎤

**Review by dakotalillie — APPROVED** [review]
LGTM 🎉

**Comment by dakotalillie** [comment]
Last commit isn't verified cuz I'm using my work laptop at the moment and haven't set up a GPG key on it yet. Will do that shortly — let me know if you'd like me to re-push the commit.

**Comment by portenez** [comment]
Given we cannot use `squash merge`, let's use the default github merge comment. 
Tools like semantic-release, usually accept those (even if they're not conventional-commits)

**Pull request merged**
